### PR TITLE
[v8r0] COMDIRAC merge

### PR DESCRIPTION
**Pull request by sfayer**
Hi,

Here is the pull request for a plain merge (no changes/fixes other than those required to get the commands to work) for discussion.

FYI @marianne013 @atsareg

Regards,
Simon

BEGINRELEASENOTES
*Interfaces
NEW: Merge COMDIRAC
ENDRELEASENOTES

### CODE REVIEW

**Comment by chaen** [comment]
Maybe I missed the discussion somewhere, but is there really a compelling reason to merge `COMDIRAC` in the main repo ? From the logical point of view, I think it makes perfect sense as a separate package...

**Comment by sfayer** [comment]
Yes, it was discussed at one of the BiLD meetings a while ago... I'm sure it'll be discussed some more at the next one.

Regards,
Simon

**Comment by fstagni** [comment]
I see from the minutes of the BiLD meeting (which unfortunately I could not attend) that there were a few discussions on this issue. 
My input is that a "plain" merge is NOT what we want: without looking in details of each of every command (someone will have to!), there are duplications (in the commands functionalities), and/or diversions. "Just new commands" is not going to be accepted.

**Comment by marianne013** [comment]
Hi Federico,
Simon and me had a long discussion with Andrei.
We *do* want a plain merge in the first instance, it does not matter if a command is duplicated (why should it?). We keep them as separate command "families" and tell users to use one or the other.
There are issues with the commands and the documentation, but once the code is in, we can go through them and fix as needed (by the looks of it, they will be at least some multi VO fixes to be done).  But this is so much easier once the code is merged, rather than do it separately. I've kind of volunteered to write a certification test - I'm just swamped with something else now.
At least at the start there will be no user base for it, as no-one knows, so it won't interfere with anyone's workflow.

**Comment by fstagni** [comment]
My only issue with "in a first instance" is that it usually ends up being "for a long time" (like many "temporary" workaround). So I would prefer to see this work of going through a fix them started already in this merge PR.

**Comment by marianne013** [comment]
I have now looked at and ran every command and fixed all the issues I found. This should now be ready for review. Merry Christmas.

File: setup.cfg
Comment (chrisburr) [inline setup.cfg]: Do we need `dunsetenv` and `dsetenv`?

File: src/DIRAC/Interfaces/Utilities/DConfigCache.py
Comment (chrisburr) [inline src/DIRAC/Interfaces/Utilities/DConfigCache.py]: To avoid leaking a file descriptor:
```suggestion
            with open(self.configCacheName, "rb") as fh:
                gConfigurationData.mergedCFG = pickle.load(fh)
```
Also should this be wrapped in `try...except` and delete the cache if loading fails (pickle isn't a "stable" format).

File: src/DIRAC/Interfaces/Utilities/DConfigCache.py
Comment (chrisburr) [inline src/DIRAC/Interfaces/Utilities/DConfigCache.py]:
```suggestion
```

File: src/DIRAC/Interfaces/Utilities/DConfigCache.py
Comment (chrisburr) [inline src/DIRAC/Interfaces/Utilities/DConfigCache.py]: https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir

File: src/DIRAC/Interfaces/Utilities/DConfigCache.py
Comment (chrisburr) [inline src/DIRAC/Interfaces/Utilities/DConfigCache.py:24]: Is this set anywhere?

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (chrisburr) [inline src/DIRAC/Interfaces/Utilities/DCommands.py:148]: Should we default to `.config/dirac` rather than littering people's home directories?

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (chrisburr) [inline src/DIRAC/Interfaces/Utilities/DCommands.py:10]: Do we need another config file (in a different format) lying around? There is already `~/.dirac.cfg`.

File: src/DIRAC/Interfaces/scripts/dinit.py
Comment (chrisburr) [inline src/DIRAC/Interfaces/scripts/dinit.py:1]: I'm not sure we need a third way to start using DIRAC

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DCommands.py:148]: DIRAC has been using ~/.dirac for a long time (easily recognizable by the '-' vs '_' in the file names :
(base) lx03:.dirac > ls
dcommands.conf                    dirac-admin-sysadmin-cli.history  dirac-configuration-cli.history
dirac_admin_sysadmin_cli.history  dirac-admin-sysadmin-.history     dsession.b49691a33e4c.44791

I'd rather not change that.

File: src/DIRAC/Interfaces/scripts/dinit.py
Comment (marianne013) [reply src/DIRAC/Interfaces/scripts/dinit.py:1]: You don't like the dinit ? As far as I understand this relates to the the caching mechanism @atsareg was so keen about.

File: src/DIRAC/Interfaces/Utilities/DConfigCache.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DConfigCache.py:24]: Hmmmm... It doesn't look that way. @atsareg  Do you know if this might be a hold over from an earlier implementation? It should probably just go.

File: src/DIRAC/Interfaces/scripts/dinit.py
Comment (marianne013) [reply src/DIRAC/Interfaces/scripts/dinit.py:1]: Though I somewhat agree that we should probably take out the proxy init stuff etc. But could I please ask for clemency and file this as an issue and I promise to fix it in pass 2 ?

File: src/DIRAC/Interfaces/Utilities/DConfigCache.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DConfigCache.py]: Fixed.

File: src/DIRAC/Interfaces/Utilities/DConfigCache.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DConfigCache.py]: Fixed

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DCommands.py:10]: dirac.cfg is autogenerated for users, while dcommands.conf wants editing. I'd prefer to quarantine anything that invites user input. I know my users :-D

File: src/DIRAC/Interfaces/Utilities/DConfigCache.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DConfigCache.py:24]: Used to be set in bashrc ?

Comment (chrisburr) [reply src/DIRAC/Interfaces/Utilities/DConfigCache.py:24]: IIRC it used to be set in the script wrappers generated by `dirac-install.py`.

File: src/DIRAC/Interfaces/scripts/dinit.py
Comment (chrisburr) [reply src/DIRAC/Interfaces/scripts/dinit.py:1]: > But could I please ask for clemency and file this as an issue and I promise to fix it in pass 2 ?

If it was targetting integration maybe but we shouldn't be randomly adding and removing stuff from production branches.

File: src/DIRAC/Interfaces/Utilities/DConfigCache.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DConfigCache.py]: True. Fixed

File: setup.cfg
Comment (marianne013) [reply setup.cfg]: Yes, it sets environment in the DIRAC session.

File: src/DIRAC/Interfaces/scripts/dinit.py
Comment (marianne013) [reply src/DIRAC/Interfaces/scripts/dinit.py:1]: It's always been there fore the COMDIRAC users, so here we just do a 1:1 transfer of functionality. We can the deprecate it in integration.

Comment (chrisburr) [reply src/DIRAC/Interfaces/scripts/dinit.py:1]: It's never been there for non-COMDIRAC users

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (fstagni) [reply src/DIRAC/Interfaces/Utilities/DCommands.py:148]: While `.config/dirac` would certainly feel more "standard", I agree with @marianne013 here. If we want to change this location, we need to do it also elsewhere in DIRAC, in a separate PR (and in integration branch).

File: docs/source/UserGuide/Tutorials/ShorthandCommands/index.rst
Comment (fstagni) [inline docs/source/UserGuide/Tutorials/ShorthandCommands/index.rst:1]: Add as first lines of this file:

```
.. _shorthand_commands:

.. set highlighting to console input/output
.. highlight:: console
```

File: docs/source/UserGuide/Tutorials/ShorthandCommands/index.rst
Comment (fstagni) [inline docs/source/UserGuide/Tutorials/ShorthandCommands/index.rst]:
```suggestion
Now edit the configuration file using your details. If you are a member of more than one VO, you might want to setup different profiles for each.
```

File: docs/source/UserGuide/commands.rst
Comment (fstagni) [inline docs/source/UserGuide/commands.rst]:
```suggestion
To be able to use these commands, please check the :ref:`shorthand_commands` section for an
```

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (fstagni) [inline src/DIRAC/Interfaces/Utilities/DCommands.py]: Would you put some order here?

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (fstagni) [inline src/DIRAC/Interfaces/Utilities/DCommands.py]: This looks quite python 2

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (fstagni) [inline src/DIRAC/Interfaces/Utilities/DCommands.py]: Why suddenly we need a new log level?

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (fstagni) [inline src/DIRAC/Interfaces/Utilities/DCommands.py:314]: These env variables...? Why? And where are they defined?

**Comment by fstagni** [comment]
I can see many small-ish details that could be adjusted (the code still feels python2-ish) but the real issues with this, IMO, are :
- duplication of functionalities
- too many explicit dependencies to "proxy stuff", while we are experimenting tokens

I almost see COMDIRAC as an alternative UI. Maybe more user-friendly than the (long-ish) list of commands that DIRAC normally provides (most of them won't be useful from a "simple user" perspective), but I don't see these new commands as something to add on top of the existing ones.

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (chrisburr) [reply src/DIRAC/Interfaces/Utilities/DCommands.py:148]: I'm pretty sure `~/.dirac` is only used by two admin-only commands to store their history (which shouldn't be in `.config`).

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (chrisburr) [reply src/DIRAC/Interfaces/Utilities/DCommands.py:10]: Do any users actually use `~/.dirac.cfg`?

File: setup.cfg
Comment (chrisburr) [reply setup.cfg]: I don't understand what that means? Sourcing the `diracosrc` sets variables and if people want more there is `export`. Why does DIRAC need a custom mechanism on top of that?

File: src/DIRAC/Interfaces/scripts/dinit.py
Comment (marianne013) [reply src/DIRAC/Interfaces/scripts/dinit.py:1]: @fstagni Could I please ask for a manager level override for this issue ? I would like to keep this for the time being for backwards compatibility and once it gets used a bit more we can streamline it.

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DCommands.py]: It seems to pass the python3 checks, so for the time being I would prefer to leave it in.

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DCommands.py]: If I rename the function "exit_with_error" (because that's what it is), would that be OK ? It's used in 6 different commands, so it's a bit of an overhead to change/remove it, for (at least from my point of view) little gain.

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DCommands.py:314]: I think the idea was that COMDIRAC has it's own ecosystem and does not interfere with anything else. Changing this now would break it for existing users; this would also be something for a more long term development. @atsareg: Any comment on this ?

File: docs/source/UserGuide/Tutorials/ShorthandCommands/index.rst
Comment (marianne013) [reply docs/source/UserGuide/Tutorials/ShorthandCommands/index.rst]: Of course I just hit "commit code" and now I have a CI test failure again, as the default commit messages are not DIRAC compliant. Sigh.

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (chrisburr) [reply src/DIRAC/Interfaces/Utilities/DCommands.py]: Nevermind Python 3, this looks quite Python 2.4 😉 (and even then it was buggy as it should have be using `try...finally`).

```suggestion
    def write(self):
        with open(self.configPath, "w") as fh:
            self.config.write(fh)
```

(The Python 3 way would be to use `pathlib`.)

**Comment by marianne013** [comment]
The flurry of recent commit is me trying to dig myself out of a github hole. Sigh.

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (fstagni) [reply src/DIRAC/Interfaces/Utilities/DCommands.py]: IMHO it is just making confusion, so please remove it. And remove also `error` at line 20.

File: src/DIRAC/Interfaces/scripts/dinit.py
Comment (fstagni) [reply src/DIRAC/Interfaces/scripts/dinit.py:1]: Let's see who's more convincing at the coming BiLD meeting.

File: docs/source/UserGuide/commands.rst
Comment (marianne013) [reply docs/source/UserGuide/commands.rst]: This update fails the checks - I don't think any of the individual subsections in the commands reference section have labels.

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DCommands.py]: Done

File: src/DIRAC/Interfaces/Utilities/DCommands.py
Comment (marianne013) [reply src/DIRAC/Interfaces/Utilities/DCommands.py]: Done

File: setup.cfg
Comment (fstagni) [reply setup.cfg]: Delete them already.

**Comment by marianne013** [comment]
@fstagni As agreed(?) during the last Bild meeting, do you have the minimum list of things we need to do before merge ?

**Comment by fstagni** [comment]
> @fstagni As agreed(?) during the last Bild meeting, do you have the minimum list of things we need to do before merge ?

I have just now created https://github.com/DIRACGrid/DIRAC/issues/6746 for 8.1.

This PR can be merged now, but the only condition is to remove *dsetenv* and *dunsetenv*.

**Comment by marianne013** [comment]
@fstagni I think we are done for v8.0. I've just done a rebase, but I realised the squalchemy fixes aren't in yet, I think this is what makes the tests fail. I guess I'll wait for them to be merged and rebase one more time ?

**Review by chrisburr — APPROVED** [review]

**Pull request merged**
